packages/components/src/number-control/index.tsx

```diff
@@ -165,13 +169,23 @@ function UnforwardedNumberControl(
 
 				if ( delta !== 0 ) {
 					delta = Math.ceil( Math.abs( delta ) ) * Math.sign( delta );
-					const distance = delta * modifier * directionModifier;
+					const distance = delta * computedStep * directionModifier;
+
+					const valueToConstrain = isValueEmpty( currentValue )
+						? baseValue
+						: ensureNumber( currentValue );
 
-					// @ts-expect-error TODO: Resolve discrepancy between `value` types in InputControl based components
-					nextState.value = constrainValue(
-						// @ts-expect-error TODO: Investigate if it's ok for currentValue to be undefined
```

Comment:
Solved this by adding a pre-computation variable (see `valueToConstrain` a few lines above), which _should_ not introduce new behavior to the component